File: stkpush/src/main/java/com/freddygenicho/mpesa/stkpush/model/Mpesa.java
package com.freddygenicho.mpesa.stkpush.model;

import android.util.Base64;
import android.util.Log;

import com.freddygenicho.mpesa.stkpush.Mode;
import com.freddygenicho.mpesa.stkpush.api.RetroClient;
import com.freddygenicho.mpesa.stkpush.api.response.STKPushResponse;
import com.freddygenicho.mpesa.stkpush.interfaces.STKListener;
import com.freddygenicho.mpesa.stkpush.interfaces.STKQueryListener;
import com.freddygenicho.mpesa.stkpush.interfaces.TokenListener;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.UnsupportedEncodingException;

import retrofit2.HttpException;
import retrofit2.Response;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

/**
 * @author Fredrick Ochieng on 02/02/2018.
 */

public class Mpesa {

    private static final String TAG = Mpesa.class.getSimpleName();

    private CompositeSubscription mCompositeSubscription;
    private String consumerKey;
    private String consumerSecret;
    private Mode mode;

    public Mpesa(String consumerKey, String consumerSecret, Mode mode) {
        this.consumerKey = consumerKey;
        this.consumerSecret = consumerSecret;
        this.mode = mode;
        this.mCompositeSubscription = new CompositeSubscription();
    }

    /**
     * returns mpesa access token
     *
     * @param tokenListener - callback listener
     */
    public void getToken(final TokenListener tokenListener) throws UnsupportedEncodingException {
        if (tokenListener == null) {
            throw new RuntimeException("Activity must implement TokenListener");
        }

        mCompositeSubscription.add(RetroClient.getApiService(mode).generateAccessToken(getAuth())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<Token>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        tokenListener.OnTokenError(e);
                    }

                    @Override
                    public void onNext(Token token) {
                        tokenListener.onTokenSuccess(token);
                    }
                }));
    }

    /**
     * generate mpesa bearer key
     *
     * @return - encoded auth string
     * @throws UnsupportedEncodingException - exception
     */
    private String getAuth() throws UnsupportedEncodingException {
        if (consumerKey.isEmpty()) {
            throw new RuntimeException("Consumer key cannot be empty");
        }

        if (consumerSecret.isEmpty()) {
            throw new RuntimeException("Consumer secret cannot be empty");
        }

        String consumerKeySecret = consumerKey + ":" + consumerSecret;
        byte[] bytes = consumerKeySecret.getBytes("ISO-8859-1");
        return "Basic " + Base64.encodeToString(bytes, Base64.NO_WRAP);
    }

    /**
     * start mpesa push stk
     *
     * @param token       - access token from mpesa
     * @param stkPush     - stk push object
     * @param stkListener - callback method
     */
    public void startStkPush(Token token, STKPush stkPush, final STKListener stkListener) {

        if (token == null) {
            throw new RuntimeException("Token cannot be null");
        }

        if (stkPush == null) {
            throw new RuntimeException("STKPush cannot be null");
        }

        if (stkListener == null) {
            throw new RuntimeException("Activity must implement TokenListener");
        }

        if (stkPush.getCallBackURL() == null) {
            throw new RuntimeException("Callback URL cannot be null");
        }

        if (stkPush.getCallBackURL().isEmpty()) {
            throw new RuntimeException("Callback URL is required");
        }

        if (stkPush.getPassword() == null) {
            throw new RuntimeException("Password can not be null");
        }

        if (stkPush.getPassword().isEmpty()) {
            throw new RuntimeException("Password is required");
        }

        mCompositeSubscription.add(RetroClient.getApiService(mode).stkPush(getAuthorization(token.getAccessToken()), stkPush)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<STKPushResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable throwable) {
                        stkListener.onError(throwable);
                    }

                    @Override
                    public void onNext(STKPushResponse stkPushResponse) {
                        stkListener.onResponse(stkPushResponse);
                    }
                }));
    }

    private String getAuthorization(String accessToken) {
        return "Bearer " + accessToken;
    }

    public void stkPushQuery(Token token, STKQuery stkQuery, final STKQueryListener stkQueryListener) {

        if (token == null) {
            throw new RuntimeException("Token cannot be null");
        }

        if (stkQuery == null) {
            throw new RuntimeException("STKQuery cannot be null");
        }

        if (stkQueryListener == null) {
            throw new RuntimeException("Activity must implement STKQueryListener");
        }

        mCompositeSubscription.add(RetroClient.getApiService(mode).stkPushQuery(getAuthorization(token.getAccessToken()), stkQuery)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<STKPushResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        stkQueryListener.onError(e);
                    }

                    @Override
                    public void onNext(STKPushResponse stkPushResponse) {
                        stkQueryListener.onResponse(stkPushResponse);
                    }
                }));
    }
}

File: stkpush/src/main/java/com/freddygenicho/mpesa/stkpush/model/Mpesa.kt
package com.freddygenicho.mpesa.stkpush.model

import android.util.Base64
import com.freddygenicho.mpesa.stkpush.Mode
import com.freddygenicho.mpesa.stkpush.api.RetroClient
import com.freddygenicho.mpesa.stkpush.api.response.STKPushResponse
import com.freddygenicho.mpesa.stkpush.interfaces.STKListener
import com.freddygenicho.mpesa.stkpush.interfaces.STKQueryListener
import com.freddygenicho.mpesa.stkpush.interfaces.TokenListener
import com.freddygenicho.mpesa.stkpush.model.Mpesa
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import java.io.UnsupportedEncodingException

/**
 * @author Fredrick Ochieng on 02/02/2018.
 */
class Mpesa(
    private val consumerKey: String,
    private val consumerSecret: String,
    private val mode: Mode
) {
    private val mCompositeSubscription: CompositeSubscription = CompositeSubscription()

    /**
     * returns mpesa access token
     *
     * @param tokenListener - callback listener
     */
    @Throws(UnsupportedEncodingException::class)
    fun getToken(tokenListener: TokenListener?) {
        if (tokenListener == null) {
            throw RuntimeException("Activity must implement TokenListener")
        }
        mCompositeSubscription.add(
            RetroClient.getApiService(mode).generateAccessToken(auth)
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribe(object : Subscriber<Token?>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        tokenListener.onTokenError(e)
                    }

                    override fun onNext(token: Token?) {
                        tokenListener.onTokenSuccess(token)
                    }
                })
        )
    }

    /**
     * generate mpesa bearer key
     *
     * @return - encoded auth string
     * @throws UnsupportedEncodingException - exception
     */
    @get:Throws(UnsupportedEncodingException::class)
    private val auth: String
        get() {
            if (consumerKey.isEmpty()) {
                throw RuntimeException("Consumer key cannot be empty")
            }
            if (consumerSecret.isEmpty()) {
                throw RuntimeException("Consumer secret cannot be empty")
            }
            val consumerKeySecret = "$consumerKey:$consumerSecret"
            val bytes = consumerKeySecret.toByteArray(charset("ISO-8859-1"))
            return "Basic " + Base64.encodeToString(bytes, Base64.NO_WRAP)
        }

    /**
     * start mpesa push stk
     *
     * @param token       - access token from mpesa
     * @param stkPush     - stk push object
     * @param stkListener - callback method
     */
    fun startStkPush(token: Token?, stkPush: STKPush?, stkListener: STKListener?) {
        if (token == null) {
            throw RuntimeException("Token cannot be null")
        }
        if (stkPush == null) {
            throw RuntimeException("STKPush cannot be null")
        }
        if (stkListener == null) {
            throw RuntimeException("Activity must implement TokenListener")
        }
        if (stkPush.callBackURL == null) {
            throw RuntimeException("Callback URL cannot be null")
        }
        if (stkPush.callBackURL!!.isEmpty()) {
            throw RuntimeException("Callback URL is required")
        }
        if (stkPush.password == null) {
            throw RuntimeException("Password can not be null")
        }
        if (stkPush.password!!.isEmpty()) {
            throw RuntimeException("Password is required")
        }
        mCompositeSubscription.add(
            RetroClient.getApiService(mode)
                .stkPush(token.accessToken?.let { getAuthorization(it) }, stkPush)
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribe(object : Subscriber<STKPushResponse?>() {
                    override fun onCompleted() {}
                    override fun onError(throwable: Throwable) {
                        stkListener.onError(throwable)
                    }

                    override fun onNext(stkPushResponse: STKPushResponse?) {
                        stkListener.onResponse(stkPushResponse)
                    }
                })
        )
    }

    private fun getAuthorization(accessToken: String): String {
        return "Bearer $accessToken"
    }

    fun stkPushQuery(token: Token?, stkQuery: STKQuery?, stkQueryListener: STKQueryListener?) {
        if (token == null) {
            throw RuntimeException("Token cannot be null")
        }
        if (stkQuery == null) {
            throw RuntimeException("STKQuery cannot be null")
        }
        if (stkQueryListener == null) {
            throw RuntimeException("Activity must implement STKQueryListener")
        }
        mCompositeSubscription.add(
            RetroClient.getApiService(mode)
                .stkPushQuery(token.accessToken?.let { getAuthorization(it) }, stkQuery)
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribe(object : Subscriber<STKPushResponse?>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        stkQueryListener.onError(e)
                    }

                    override fun onNext(stkPushResponse: STKPushResponse?) {
                        stkQueryListener.onResponse(stkPushResponse)
                    }
                })
        )
    }

    companion object {
        private val TAG = Mpesa::class.java.simpleName
    }
}